src/cast.c

```diff
@@ -50,24 +50,27 @@ r_obj* vec_cast_opts(const struct cast_opts* opts) {
     stop_scalar_type(to, to_arg, opts->call);
   }
 
-  if (has_dim(x) || has_dim(to)) {
-    return vec_cast_dispatch_s3(opts);
-  }
-
   r_obj* out = r_null;
   bool lossy = false;
 
   if (to_type == VCTRS_TYPE_s3 || x_type == VCTRS_TYPE_s3) {
-    out = vec_cast_dispatch_native(opts, x_type, to_type, &lossy);
+    out = KEEP(vec_cast_dispatch_native(opts, x_type, to_type, &lossy));
   } else {
-    out = vec_cast_switch_native(opts, x_type, to_type, &lossy);
+    out = KEEP(vec_cast_switch_native(opts, x_type, to_type, &lossy));
   }
 
   if (lossy || out == r_null) {
+    // This broadcasts dimensions too
+    FREE(1);
     return vec_cast_dispatch_s3(opts);
```

Comment:
The idea here is that if you have a custom S3 type, and you expect that it can have dimensions, then you have to handle that yourself? Just checking

Reply:
I think this is just falling back to the R implementations because we have been too lazy/busy to port the lossy cast paths to C.

Reply:
Isn't this the standard S3 dispatch path too? i.e. if `out == r_null` we never hit a class we handle natively

Reply:
ah maybe. But yes the intention was only to support dimensions on base classes.

It seems like it'd be too large a change to handle dimensions for S3 classes? I'm not sure how this would interact with classes like in rray.

Reply:
> It seems like it'd be too large a change to handle dimensions for S3 classes?

Yea my instinct is to not handle dimensions automatically for other S3 classes. Too magical and error prone I think